### PR TITLE
Remove sneaky fredkin decomposition

### PR DESCRIPTION
**Pull request by mpharrigan**
This should be a gate/bloq

### CODE REVIEW

**Comment by mpharrigan** [comment]
receipts: https://github.com/quantumlib/Qualtran/pull/393#discussion_r1357527018

File: qualtran/bloqs/basic_gates/swap.py [hunk 1]
```diff
@@ -307,7 +306,7 @@ def wire_symbol(self, soq: 'Soquet') -> 'WireSymbol':
             return Circle(filled=True)
 
     def _t_complexity_(self) -> TComplexity:
-        return TComplexity(t=7 * self.bitsize, clifford=10 * self.bitsize)
+        return TComplexity(t=7 * self.bitsize, clifford=14 * self.bitsize)
```
Comment (tanujkhattar) [inline qualtran/bloqs/basic_gates/swap.py:309]: Why this change? You decompose into `self.bitsize` `TwoBitCSwap`s, each of which returns `TComplexity(t=7, clifford=10)`

File: qualtran/bloqs/basic_gates/swap.py [hunk 2]
```diff
@@ -159,10 +159,9 @@ def _t_complexity_(self) -> 'TComplexity':
         """The t complexity.
 
         References:
-            [An algorithm for the T-count](https://arxiv.org/abs/1308.4134). Gosset et. al. 2013.
-            Figure 5.2.
+            [An algorithm for the T-count](https://arxiv.org/abs/1308.4134).
+            Gosset et. al. 2013. Figure 5.2.
         """
-        # https://arxiv.org/abs/1308.4134
         return TComplexity(t=7, clifford=10)
```
Comment (tanujkhattar) [inline qualtran/bloqs/basic_gates/swap.py:165]: Can you actually add the decomposition that was deleted from `_fredkin` here? 

```
    yield [cirq.CNOT(t2, t1)]
    yield [cirq.CNOT(c, t1), cirq.H(t2)]
    yield [cirq.T(c), cirq.T(t1) ** -1, cirq.T(t2)]
    yield [cirq.CNOT(t2, t1)]
    yield [cirq.CNOT(c, t2), cirq.T(t1)]
    yield [cirq.CNOT(c, t1), cirq.T(t2) ** -1]
    yield [cirq.T(t1) ** -1, cirq.CNOT(c, t2)]
    yield [cirq.CNOT(t2, t1)]
    yield [cirq.T(t1), cirq.H(t2)]
    yield [cirq.CNOT(t2, t1)]
```

File: qualtran/cirq_interop/t_complexity_protocol_test.py [hunk 3]
```diff
@@ -130,7 +130,7 @@ def test_gates():
     assert t_complexity(And()) == TComplexity(t=4, clifford=9)
     assert t_complexity(And() ** -1) == TComplexity(clifford=4)
 
-    assert t_complexity(cirq.FREDKIN) == TComplexity(t=7, clifford=10)
+    assert t_complexity(cirq.FREDKIN) == TComplexity(t=7, clifford=14)
```
Comment (tanujkhattar) [inline qualtran/cirq_interop/t_complexity_protocol_test.py:133]: Why this change?

**Review by tanujkhattar — CHANGES_REQUESTED** [review]

File: qualtran/cirq_interop/t_complexity_protocol_test.py (hunk 3 above)
Comment (mpharrigan) [reply qualtran/cirq_interop/t_complexity_protocol_test.py:133]: that's what it returns 🤷‍♀️

Comment (tanujkhattar) [reply qualtran/cirq_interop/t_complexity_protocol_test.py:133]: Okay sure, I missed that part that we are directly calling `cirq.FREDKIN`.

We should add another test somewhere (if it doesn't exist already) that 

```
assert t_complexity(CSwap(1)) == TComplexity(t=7, clifford=10)
assert t_complexity(TwoBitCSwap()) == TComplexity(t=7, clifford=10)
```

File: qualtran/bloqs/basic_gates/swap.py (hunk 1 above)
Comment (mpharrigan) [reply qualtran/bloqs/basic_gates/swap.py:309]: `test_t_complexity_cswap` fails if it's set to 10

Comment (mpharrigan) [reply qualtran/bloqs/basic_gates/swap.py:309]: it's because the test utility will go via `to_cirq_circuit` which uses `TwoBitCSwap.as_cirq_op()` to give `cirq.FREDKIN` which has the cliffords we deign not to include

Comment (mpharrigan) [reply qualtran/bloqs/basic_gates/swap.py:309]: changed the offending test

File: qualtran/bloqs/basic_gates/swap.py (hunk 2 above)
Comment (mpharrigan) [reply qualtran/bloqs/basic_gates/swap.py:165]: added

File: qualtran/cirq_interop/t_complexity_protocol_test.py (hunk 3 above)
Comment (mpharrigan) [reply qualtran/cirq_interop/t_complexity_protocol_test.py:133]: added

**Comment by mpharrigan** [comment]
@tanujkhattar ptal

File: qualtran/bloqs/swap_network/cswap_approx_test.py [hunk 4]
```diff
@@ -53,8 +54,15 @@ def get_t_count_and_clifford(bc: Dict[Bloq, int]) -> Tuple[int, int]:
 
 
 @pytest.mark.parametrize("n", [*range(1, 6)])
-def test_t_complexity(n):
-    cq_testing.assert_decompose_is_consistent_with_t_complexity(CSwap(n))
+def test_t_complexity_cswap(n):
+    # Note: if you use the cirq decompose protocol, clifford counts won't
+    # match because cirq.FREDKIN has 14 cliffords per two-bit-cswap.
+    cswap = CSwap(n)
+    assert t_complexity(cswap) == t_complexity(cswap.decompose_bloq())
```
Comment (tanujkhattar) [inline qualtran/bloqs/swap_network/cswap_approx_test.py:61]: @mpharrigan This is happening because `TwoBitCSwap` has both `as_cirq_op` and `decompose_from_registers` specified.  Because `TwoBitCSwap` is no longer an atomic gate; we should delete the `as_cirq_op` implementation from it. Once that is done, `assert_decompose_is_consistent_with_t_complexity` will pass here. 

I think it would be surprising for any users using the "compilation to cirq" strategy to get a circuit with a different T-count as what they expected using the bloq analysis. Note that the circuit structure in cirq style compilation can still be preserved because decomposing a `CSwap` now gives `BloqAsCirqGate(TwoBitCSwap())` instead of a `cirq.FREDKIN`; which is fine I think and works for all protocols. 

I tried making the change locally and it results in failure of only two tests (+ 2 unrelated failures due to https://github.com/quantumlib/Qualtran/pull/826) ; both of which should be updated / removed because we no longer want to have a 1:1 correspondence between `TwoBitCSwap` and `cirq.FREDKIN` 
```
FAILED qualtran/bloqs/basic_gates/swap_test.py::test_two_bit_cswap - assert cirq.FrozenCi...2)),\n    ),\n]) == cirq.FrozenCi...2)),\n    ),\n])
FAILED qualtran/bloqs/basic_gates/swap_test.py::test_cswap_cirq_decomp - AssertionError: Circuit's text diagram differs from the desired diagram.
```

What do you think?
```suggestion
def test_t_complexity_cswap(n):
    cq_testing.assert_decompose_is_consistent_with_t_complexity(CSwap(n))
```

Comment (mpharrigan) [reply qualtran/bloqs/swap_network/cswap_approx_test.py]: done
